Feature/49 version command

JR-Mitchell opened the pull request:
### What?
Adds `version.py`, which includes:
- Variables for setting the current semantic version of the bot
- A `Version` cog for checking the currently running version

### Why?
Resolves #49

JR-Mitchell committed:
Adds basic version cog

JR-Mitchell committed:
Adds variables and method for semantic versioning in version.py

JR-Mitchell committed:
Fixes linting issues

JR-Mitchell committed:
Fixes reversed logic

fjebaker commented on src/cogs/version.py:
```diff
@@ -0,0 +1,30 @@
+MAJOR_VERSION = 0
+MINOR_VERSION = 1
+PATCH_VERSION = 0
+RELEASE_CANDIDATE = 1
```
I love the optimisim that there _would ever be a release_.

fjebaker commented:
Principally this is fine but I have a couple of gripes:
- The version info is hidden in a cog. Would the global variables not be better positioned in `ebot/__init__.py`, as is more standard?
- Semver is good but do we really need RC? If I make a change, do I bump patch or RC? When would I bump one over the other? Cus we don't have a release schedule, I think RC can be dropped and we just always bump patch unless breaking.

Then also
- We need to remember to bump version info else make a workflow at some point that ensures code changes also bump minor, but this is a common problem with every project so don't fret about it in this PR. Just something to think about.

fjebaker commented on src/cogs/version.py:
```diff
@@ -0,0 +1,30 @@
+class Version(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+        self.logging = logging.getLogger(__name__)
```
Can you add a log statement that just logs the current version to `info` for easier debug?

JR-Mitchell replied:
Added as of [f833022](https://github.com/fjebaker/e-bot/pull/50/commits/f83302236594350d4c61c55d38f1ad0eefd152e6) - I also made the command text more descriptive :)

JR-Mitchell commented:
> Semver is good but do we really need RC? If I make a change, do I bump patch or RC? When would I bump one over the other? Cus we don't have a release schedule, I think RC can be dropped and we just always bump patch unless breaking.

@fjebaker 
One thing I would like to see in the future is distinct `main` and `develop` branches for this project.
Then we could perhaps have a test server and bot instance, where we can perform various tests on the bot after merging work, distinct from any servers where the bot is actually used.

The way this would work is:

- Version `{a}.{b}.{c}` exists in the `main` branch, which is up to date with the `develop` branch and vice versa
- We either make a new feature, and so make a PR for the feature to `develop` which changes the code to `{a}.{b+1}.{0}-rc1`, or we notice a bug in `main`, and so make a PR for the fix to `develop` which changes the code to `{a}.{b}.{c+1}-rc1`
- Once the PR is merged, the new RC version is deployed to the testing instance, and we can run whatever tests we deem necessary (e.g simple stress test on the command for a new cog, try to replicate a bug in order to demonstrate that it's fixed, or regression tests around the changed code)
- If a test fails, we need another PR, and bump the version to `-rc2`, and so on until we pass
- If all tests pass, we make a commit to `develop` which removes the RC version, and then sync to `main` (and tag with the version number)

I know this is thinking pretty far ahead and we might not ever get round to doing this - but, imo, there's nothing wrong with having the constant and logic for it ready there so that we can use it if and when we get to that stage, right?

JR-Mitchell committed:
Adds logging and more descriptive message to Version command

JR-Mitchell committed:
Reduces RC version to 0 for now

fjebaker commented:
@JR-Mitchell I think the system you outline is a perfectly valid and functional, _but_ I believe it (currently) beyond the scope of the bot. The bot only exists on one server, and it's buggy and not always available (credit to my grandmother).

I think we just treat the `main` branch as the develop branch, and trial features directly in production with the pi pulling and reloading whenever changes are made. Obviously this is a terrible idea for any project, but then, this is a terrible bot, and the flower is only as good as the mud it grew from.

fjebaker commented:
So I still think we get rid of the RC part, and just bump patch.

JR-Mitchell committed:
Moves the version numbers to their own top-level declaration file

JR-Mitchell committed:
Fixes lint issues in version.py

JR-Mitchell commented:
@fjebaker I've moved the version info to a `src/eversion.py` file.
I've kept the RC version there for the possibility that we improve our flow in the future.
For now we can safely ignore it and just bump patch, but I see no reason to remove it from the codebase altogether.
I've added a comment next to it saying that it's currently not in use, but may be in the future.
You happy for me to merge?

JR-Mitchell committed:
Changed the value of eversion.RELEASE_CANDIDATE from 0 to None
